### DISCUSSION TITLE
Beartype and pytest for the win!

### DISCUSSION THREAD

**Discussion by Jasha10** [beartype/beartype#199] [General]
The thesis here is that pytest tests and fixtures can (and should!) be decorated with `@beartype`.
Doing so makes your tests robust against mis-typed fixtures and mis-typed `@mark.parametrize` decorators.

## Motivation
Sometimes using the pytest framework is risky:

```python
from pytest import fixture

@fixture
def foo_int() -> int:
    return 123

def test_foo(foo_int: str) -> None:
    assert foo_int == "abc"
```
It's easy to see that the above test will fail since `123 != "abc"`. But can you spot the type error?
```text
$ pytest test_foo.py
...
FAILED test_foo.py::test_foo - AssertionError: assert 123 == 'abc'
1 failed in 0.05s
```
Running static checkers like `mypy` and `pyright` on the above code will *not* produce any errors.
... But the `foo_int` argument passed to the `test_foo` function actually doesn't have type `str`,
despite what the type annotation `foo_int: str` would lead you to believe.
I guess static typing can't solve everything.

## Enter beartype!

With the addition of a decorator, you can ensure that your mis-typed code will crash spectacularly:
```python
from beartype import beartype
from pytest import fixture

@fixture
def foo_int() -> int:
    return 123

@beartype  # Add decorator here!
def test_foo(foo_int: str) -> None:
    assert foo_int == "abc"
```
```text
$ pytest test_foo.py -q
...
E   beartype.roar.BeartypeCallHintParamViolation: @beartyped test_foo.test_foo() parameter foo_int=123 violates type hint <class 'str'>, as int 123 not instance of str.

<@beartype(test_foo.test_foo) at 0x7f18eac17130>:22: BeartypeCallHintParamViolation
====================== short test summary info =======================
FAILED test_foo.py::test_foo - beartype.roar.BeartypeCallHintParamViolation: @beartyped test_foo...
1 failed in 0.10s
```
The test doesn't run! We are immediately given a bear-typed slap on the wrist.
The above diagnostic info is enough to piece together that the signature of `test_foo` is mistyped.

What advantages are there to using `@beartype` with pytest? Why should I care whether my test fails with a `BeartypeCallHintParamViolation` or with a standard pytest `AssertionError` as in the first example above?
Using `@beartype`, you can follow the Single Responsibility Principle in your test code. Your test functions' body needs only to worry about test logic, not about verifying that the fixtures or `parametrize`'d arguments are correctly-typed.

## Recipes:

Beartype works with pytest tests:
```python
from beartype import beartype
@beartype  # ensure `foo_int` *actually* has type `str`
def test_foo(foo_int: str) -> None:
    assert foo_int == "abc"
```
Beartype works with pytest fixtures:
```python
from beartype import beartype
from pytest import fixture
@fixture
@beartype  # ensure the `foo_int` fixture *actually* returns an `int`
def foo_int() -> int:
    return 123

# # Caveat lector: If your fixture yields a value, decoration with `@beartype` might not work.
# # Pytest doesn't handle this properly:
# @fixture
# @beartype
# def foo_int() -> typing.Iterable[int]:
#     yield 123
```

Beartype can make sure your `@mark.parametrize` arguments are up to snuff:
```python
from beartype import beartype
from pytest import mark
@mark.parametrize("arg", [1, 2, "three"])
@beartype  # Ensure that `arg` *actually* has type `int`
def test_small_ints(arg: int) -> None:
    assert arg < 4
```

**Comment by leycec** [comment]
**Excellent expose!** This would be fantastic fodder for our monolithic `README.rst` that increasingly frightens even me. Specifically:

* Let's shamelessly copy-paste all three of the `pytest` recipes you've so graciously donated to the QA world above into a new `PYTEST` subsection of [our existing *Cheatsheet*](https://github.com/beartype/beartype#cheatsheet).
* Let's shamelessly copy-paste your *wonderfully* well-written introduction into a new FAQ entry on why (*"because you'd be a blatant fool not to!"*), when (*always. just do it always.*), and how (which is kinda non-trivial, because order of decorations is significant) to integrate @beartype with `pytest`.

To remind myself to do everything above, [I've transitioned this discussion into a full-fledged issue](https://github.com/beartype/beartype/issues/200). Thanks so much for your superbly helpful write-up, @Jasha10! I especially :heart_eyes: reading soothing words like this:

> I guess static typing can't solve everything.

![nods approval](https://media.tenor.com/dB61WWrIAY4AAAAS/gif.gif)
